src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -1529,32 +1564,20 @@ private static unsafe nuint NarrowUtf16ToAscii_Sse2(char* pUtf16Buffer, byte* pA
             {
                 // In a loop, perform two unaligned reads, narrow to a single vector, then aligned write one vector.
 
-                utf16VectorFirst = Sse2.LoadVector128((short*)pUtf16Buffer + currentOffsetInElements); // unaligned load
-                Vector128<short> utf16VectorSecond = Sse2.LoadVector128((short*)pUtf16Buffer + currentOffsetInElements + SizeOfVector128 / sizeof(short)); // unaligned load
-                Vector128<short> combinedVector = Sse2.Or(utf16VectorFirst, utf16VectorSecond);
+                utf16VectorFirst = Vector128.LoadUnsafe(ref utf16Buffer, currentOffsetInElements); // unaligned load
+                Vector128<ushort> utf16VectorSecond = Vector128.LoadUnsafe(ref utf16Buffer, currentOffsetInElements + SizeOfVector128 / sizeof(short)); // unaligned load
+                Vector128<ushort> combinedVector = utf16VectorFirst | utf16VectorSecond;
 
-                // See comments in GetIndexOfFirstNonAsciiChar_Sse2 for information about how this works.
-                if (Sse41.IsSupported)
-                {
-                    if (!Sse41.TestZ(combinedVector, asciiMaskForTestZ))
-                    {
-                        goto FoundNonAsciiDataInLoop;
-                    }
-                }
-                else
+                if (VectorContainsNonAsciiChar(combinedVector))
                 {
-                    if ((Sse2.MoveMask(Sse2.AddSaturate(combinedVector.AsUInt16(), asciiMaskForAddSaturate).AsByte()) & NonAsciiDataSeenMask) != 0)
-                    {
-                        goto FoundNonAsciiDataInLoop;
-                    }
+                    goto FoundNonAsciiDataInLoop;
                 }
 
                 // Build up the ASCII vector and perform the store.
 
-                asciiVector = Sse2.PackUnsignedSaturate(utf16VectorFirst, utf16VectorSecond);
-
                 Debug.Assert(((nuint)pAsciiBuffer + currentOffsetInElements) % SizeOfVector128 == 0, "Write should be aligned.");
-                Sse2.StoreAligned(pAsciiBuffer + currentOffsetInElements, asciiVector); // aligned
+                asciiVector = ExtractAsciiVector(utf16VectorFirst, utf16VectorSecond);
+                asciiVector.StoreAligned(pAsciiBuffer + currentOffsetInElements); // aligned store
```

Comment:
@tannergooding, we can improve `Vector128.StoredAligned` on AArch64.
Currently, it emits code for `AccessViolationException`.

- With `Vector128.StoreUnsafe()`
```
    st1     {v16.16b}, [x2]
```
[Full dump](https://gist.github.com/SwapnilGaikwad/62ef6ef76e49843b7b7c8ccb9be4286b)

- With `Vector128.StoreAligned()`
```
    add     x5, x1, x4
    tst     w5, #15
    bne     G_M20155_IG11
    str     q16, [x5]

G_M20155_IG11:
    movz    x0, #0xd1ffab1e
    movk    x0, #0xd1ffab1e LSL #16
    movk    x0, #0xd1ffab1e LSL #32
    bl      CORINFO_HELP_NEWSFAST
    mov     x19, x0
    movz    x1, #0xd1ffab1e      // code for System.AccessViolationException:.ctor
    movk    x1, #0xd1ffab1e LSL #16
    movk    x1, #0xd1ffab1e LSL #32
    ldr     x1, [x1]
    blr     x1
    mov     x0, x19
    bl      CORINFO_HELP_THROW
    brk_unix #0
```
[Full dump](https://gist.github.com/SwapnilGaikwad/9e669bff00bb3dfa53fb8e3553b2dd22)

The aligned stores take majority of the gains from AArch64.
```
|   Method |        Job |                                                                                          Toolchain | size | encName |      Mean |    Error |   StdDev |    Median |       Min |       Max | Ratio | MannWhitney(2%) |  Gen 0 | Allocated | Alloc Ratio |
|--------- |----------- |--------------------------------------------------------------------------------------------------- |----- |-------- |----------:|---------:|---------:|----------:|----------:|----------:|------:|---------------- |-------:|----------:|------------:|
| GetBytes | Job-PDGOEQ | /main-artifacts/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii |  26.12 ns | 0.028 ns | 0.024 ns |  26.12 ns |  26.09 ns |  26.17 ns |  1.00 |            Base | 0.0024 |      40 B |        1.00 |
| GetBytes | Job-DPIHLX |  /store-aligned/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii |  26.31 ns | 0.017 ns | 0.013 ns |  26.30 ns |  26.30 ns |  26.34 ns |  1.01 |            Same | 0.0023 |      40 B |        1.00 |
| GetBytes | Job-WKIBWN |   /store-unsafe/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii |  26.62 ns | 0.037 ns | 0.033 ns |  26.61 ns |  26.58 ns |  26.71 ns |  1.02 |            Same | 0.0024 |      40 B |        1.00 |
|          |            |                                                                                                    |      |         |           |          |          |           |           |           |       |                 |        |           |             |
| GetBytes | Job-PDGOEQ | /main-artifacts/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii | 103.82 ns | 0.531 ns | 0.414 ns | 103.94 ns | 102.51 ns | 103.98 ns |  1.00 |            Base | 0.0318 |     536 B |        1.00 |
| GetBytes | Job-DPIHLX |  /store-aligned/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii | 101.57 ns | 0.186 ns | 0.145 ns | 101.60 ns | 101.28 ns | 101.73 ns |  0.98 |          Faster | 0.0318 |     536 B |        1.00 |
| GetBytes | Job-WKIBWN |   /store-unsafe/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii |  96.15 ns | 0.045 ns | 0.035 ns |  96.14 ns |  96.10 ns |  96.21 ns |  0.93 |          Faster | 0.0319 |     536 B |        1.00 |
|          |            |                                                                                                    |      |         |           |          |          |           |           |           |       |                 |        |           |             |
| GetBytes | Job-PDGOEQ | /main-artifacts/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun | 1024 |   ascii | 181.55 ns | 0.112 ns | 0.094 ns | 181.53 ns | 181.42 ns | 181.72 ns |  1.00 |            Base | 0.0624 |    1048 B |        1.00 |
| GetBytes | Job-DPIHLX |  /store-aligned/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun | 1024 |   ascii | 178.45 ns | 0.390 ns | 0.326 ns | 178.33 ns | 177.98 ns | 178.96 ns |  0.98 |            Same | 0.0625 |    1048 B |        1.00 |
| GetBytes | Job-WKIBWN |   /store-unsafe/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun | 1024 |   ascii | 163.82 ns | 0.317 ns | 0.264 ns | 163.67 ns | 163.61 ns | 164.40 ns |  0.90 |          Faster | 0.0625 |    1048 B |        1.00 |
```

Contrary, using `Vector128.StoreUnsafe()` in a loop regresses x86 version significantly (by additional 14% relative to the HEAD).
We expect a few nano-seconds hit from the unnecessary moves resulting from switching to the vector API for load/stores (mentioned [above](https://github.com/dotnet/runtime/pull/70080#discussion_r902478514) )

Note: Following numbers are not from benchmarking friendly machine, so may be noisier.
```
|   Method |        Job |                                                                                        Toolchain | size | encName |      Mean |    Error |   StdDev |    Median |       Min |       Max | Ratio | MannWhitney(2%) |  Gen 0 | Allocated | Alloc Ratio |
|--------- |----------- |------------------------------------------------------------------------------------------------- |----- |-------- |----------:|---------:|---------:|----------:|----------:|----------:|------:|---------------- |-------:|----------:|------------:|
| GetBytes | Job-PZMRCU | /main-artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii |  25.01 ns | 0.243 ns | 0.216 ns |  24.89 ns |  24.87 ns |  25.49 ns |  1.00 |            Base | 0.0040 |      40 B |        1.00 |
| GetBytes | Job-YQVKBT |  /store-aligned/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii |  25.70 ns | 0.228 ns | 0.213 ns |  25.57 ns |  25.54 ns |  26.13 ns |  1.03 |          Slower | 0.0039 |      40 B |        1.00 |
| GetBytes | Job-GIVNDK |   /store-unsafe/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii |  25.84 ns | 0.222 ns | 0.208 ns |  25.78 ns |  25.59 ns |  26.27 ns |  1.03 |          Slower | 0.0039 |      40 B |        1.00 |
|          |            |                                                                                                  |      |         |           |          |          |           |           |           |       |                 |        |           |             |
| GetBytes | Job-PZMRCU | /main-artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii |  83.06 ns | 0.756 ns | 0.631 ns |  83.29 ns |  81.75 ns |  83.76 ns |  1.00 |            Base | 0.0529 |     536 B |        1.00 |
| GetBytes | Job-YQVKBT |  /store-aligned/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii |  85.93 ns | 0.670 ns | 0.594 ns |  85.68 ns |  85.42 ns |  87.28 ns |  1.03 |          Slower | 0.0532 |     536 B |        1.00 |
| GetBytes | Job-GIVNDK |   /store-unsafe/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii |  97.49 ns | 0.757 ns | 0.709 ns |  97.31 ns |  96.66 ns |  98.95 ns |  1.17 |          Slower | 0.0531 |     536 B |        1.00 |
|          |            |                                                                                                  |      |         |           |          |          |           |           |           |       |                 |        |           |             |
| GetBytes | Job-PZMRCU | /main-artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun | 1024 |   ascii | 144.68 ns | 0.401 ns | 0.335 ns | 144.60 ns | 144.29 ns | 145.46 ns |  1.00 |            Base | 0.1040 |    1048 B |        1.00 |
| GetBytes | Job-YQVKBT |  /store-aligned/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun | 1024 |   ascii | 149.11 ns | 1.161 ns | 1.029 ns | 148.73 ns | 148.19 ns | 151.40 ns |  1.03 |          Slower | 0.1036 |    1048 B |        1.00 |
| GetBytes | Job-GIVNDK |   /store-unsafe/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun | 1024 |   ascii | 171.57 ns | 1.315 ns | 1.230 ns | 170.79 ns | 170.38 ns | 173.73 ns |  1.19 |          Slower | 0.1037 |    1048 B |        1.00 |
```

Thus, this patch uses `StoreUnsafe` to avoid slowing down the existing code. We can create an issue to fix this performance issue on AArch64.

Reply:
We can, and should, just be using `StoreUnsafe` on both.

For modern CPUs on x86/x64, the instructions all support unaligned loads and so `StoreAligned` has limitations on when it can be folded (e.g. `T1` only, etc). `StoreUnaligned` does not have these limitations and doesn't imply other checks.

It's better to process the data "as if it was pinned" and optimistically operate on aligned chunks. In the normal case, the GC won't interrupt and move the data and everything is "ok". In the worst case, the GC interrupts, moves the data, and it's no longer aligned but since you're doing unaligned loads/stores, the algorithm keeps working anyways and you just pay the minor cost.

This works well with actual pinned data and unpinned data where you just have the `byref`. The latter ends up playing better with the GC as well, so we really should prefer it for most of our code.

Reply:
Thanks for your comments. Now switched to `StoreUnsafe`. Marking this conversation resolved.